docs/DesignDocs/Enabledebugmalloc.md

```diff
@@ -0,0 +1,289 @@
+tracking state cumulatively.
+
+```
+      stop()  |--------------|  start() |--------------|  start()
+   |--------->|              |--------->|              |----------|
+   |          | not tracking |          |   tracking   |          |
+   |----------|              |<---------|              |<---------|
+              |--------------|  stop()  |--------------|
+```
+
+- oe_debug_malloc_tracking_start() and oe_debug_malloc_tracking_stop() can be
+called many times:
+
+```
+  |----code1--|-----------|----code2--|--|----code3--|
+  start1->stop1->report1->start2->stop2->start3->stop3->report3
+```
+
+In this case report1 only covers code1, and report3 covers code1, code2 and
+code3.
+
+Please also note that these public APIs only affect the result of local
+detection. The result of global detection is independent of these APIs. While
+the calling of these APIs follow the rules, the result of local detection is
+supposed to be the subset of result of global detection.
+
+# Implementation
+
+The debug malloc employs a linked list to make a record of all memory operations.
+So, for the local detection, we can make use of the fulfilled record system.
+
+```c
+struct header
+{
+    /* Contains HEADER_MAGIC1 */
+    uint64_t magic1;
+
+    /* Headers are kept on a doubly-linked list */
+    header_t* next;
+    header_t* prev;
+
+    /* The alignment passed to memalign() or zero */
+    uint64_t alignment;
+
+    /* Size of user memory */
+    size_t size;
+
+    /* Return addresses obtained by oe_backtrace() */
+    void* addrs[OE_BACKTRACE_MAX];
+    uint64_t num_addrs;
+
+    /* Option if current object is tracked */
+    bool local_tracking;
+
+    /* Padding to make header a multiple of 16 */
+    uint8_t padding[7];
+
+    /* Contains HEADER_MAGIC2 */
+    uint64_t magic2;
+
+    /* User data */
+    uint8_t data[];
+};
+```
+
+We also add a variable
+```c
+bool oe_use_debug_malloc_tracking = false;
+```
+which will be changed by oe_debug_malloc_tracking_start() and
+oe_debug_malloc_tracking_stop(). While doing the allocation, the objective's
+`local_tracking` is marked as `oe_use_debug_malloc_tracking`, then
+oe_debug_malloc_tracking_report() can get the number of all the not yet freed
+objects.
+
+# Sample
+
+```c
+#include <openenclave/corelibc/stdlib.h>
+#include <openenclave/debugmalloc.h>
+#include <stdio.h>
+
+// Include the trusted debugmalloc header that is generated
+// during the build. This file is generated by calling the
+// sdk tool oeedger8r against the debugmalloc.edl file.
+#include "debugmalloc_t.h"
+
+// This is the function that the host calls. It prints
+// a message in the enclave before calling back out to
+// the host to print a message from there too.
+void enclave_hello()
+{
+    void* buf1, buf2, buf3;
+    uint64_t num;
+    char* report;
+
+    oe_debug_malloc_tracking_start();
+    void* buf1 = oe_malloc(4096);
+    oe_debug_malloc_tracking_stop();
+    oe_debug_malloc_tracking_report(&num, &report);
```

Comment:
Caller should check the return for this since failed allocations can fail this function resulting in null report pointer.

Reply:
Updated, thanks!